frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -60,41 +59,40 @@ public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNot
 
   @Override
   public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
-    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+    return Optional.ofNullable(placesByTripIdByPlaceId.get(tripId))
+        .map(placesMap -> placesMap.get(placeId));
   }
 
   @Override
   public boolean updateUserMarkOrAddPlaceVisit(
       PlaceVisitModel placeVisit, PlaceVisitModel.UserMark newStatus) {
-    synchronized (storage) {
-      Map<String, PlaceVisitModel> placesMap = storage.get(placeVisit.tripId());
-      if (placesMap == null) {
-        Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
-        newPlaceMap.put(placeVisit.placeId(), placeVisit);
-        return false;
-      }
-      if (placesMap.get(placeVisit.placeId()) != null) {
-        placesMap.put(placeVisit.placeId(),
-            placesMap.get(placeVisit.placeId()).toBuilder().setUserMark(newStatus).build());
-        return true;
+    AtomicBoolean alreadyInStorage = new AtomicBoolean(false);
+
+    Map<String, PlaceVisitModel> placesMap = placesByTripIdByPlaceId.computeIfAbsent(
+        placeVisit.tripId(), (tripKey) -> new ConcurrentHashMap<>());
+
+    placesMap.compute(placeVisit.placeId(), (placeKey, place) -> {
+      if (place != null) {
+        PlaceVisitModel updatedPlace = place.toBuilder().setUserMark(newStatus).build();
+        alreadyInStorage.set(true);
+        return updatedPlace;
```

Comment:
I'm confused, how can you return a PlaceVisitModel when the return type is a boolean

Reply:
the compute function basically just updates/adds an entry in the hashmap so the lambda function that is passed into the compute function has to return the object that you want to store as the value - it won't return from the overall function